File: jsimple-oauth/src/main/java/jsimple/oauth/extractors/HeaderExtractorImpl.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2012-2015 Microsoft Mobile.  All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *
 * This file is based on or incorporates material from the Scribe Java OAuth
 * library https://github.com/fernandezpablo85/scribe-java (collectively, "Third
 * Party Code"). Microsoft Mobile is not the original author of the Third Party
 * Code.
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package jsimple.oauth.extractors;

import jsimple.oauth.exceptions.OAuthParametersMissingException;
import jsimple.oauth.model.OAuthRequest;
import jsimple.oauth.utils.OAuthEncoder;
import jsimple.util.Map;

/**
 * Default implementation of {@link HeaderExtractor}. Conforms to OAuth 1.0a
 *
 * @author Pablo Fernandez
 */
public class HeaderExtractorImpl implements HeaderExtractor {
    private static final String PARAM_SEPARATOR = ", ";
    private static final String PREAMBLE = "OAuth ";

    /**
     * {@inheritDoc}
     */
    public String extract(OAuthRequest request) {
        checkPreconditions(request);
        Map<String, String> parameters = request.getOauthParameters();
        StringBuilder header = new StringBuilder(parameters.size() * 20);
        header.append(PREAMBLE);
        for (String key : parameters.keySet()) {
            if (header.length() > PREAMBLE.length())
                header.append(PARAM_SEPARATOR);
            header.append(key + "=\"" + OAuthEncoder.encode(parameters.get(key)) + "\"");
        }
        return header.toString();
    }

    private void checkPreconditions(OAuthRequest request) {
        if (request.getOauthParameters() == null || request.getOauthParameters().size() <= 0)
            throw new OAuthParametersMissingException(request);
    }
}

File: jsimple-oauth/c#/jsimple/oauth/extractors/HeaderExtractorImpl.cs
using System.Collections.Generic;
using System.Text;

namespace jsimple.oauth.extractors {

    using OAuthParametersMissingException = jsimple.oauth.exceptions.OAuthParametersMissingException;
    using OAuthRequest = jsimple.oauth.model.OAuthRequest;
    using OAuthEncoder = jsimple.oauth.utils.OAuthEncoder;

    /// <summary>
    /// Default implementation of <seealso cref="HeaderExtractor"/>. Conforms to OAuth 1.0a
    /// 
    /// @author Pablo Fernandez
    /// </summary>
    public class HeaderExtractorImpl : HeaderExtractor {
        private const string PARAM_SEPARATOR = ", ";
        private const string PREAMBLE = "OAuth ";

        /// <summary>
        /// {@inheritDoc}
        /// </summary>
        public virtual string extract(OAuthRequest request) {
            checkPreconditions(request);
            IDictionary<string, string> parameters = request.OauthParameters;
            StringBuilder header = new StringBuilder(parameters.Count * 20);
            header.Append(PREAMBLE);
            foreach (string key in parameters.Keys) {
                if (header.Length > PREAMBLE.Length)
                    header.Append(PARAM_SEPARATOR);
                header.Append(key + "=\"" + OAuthEncoder.encode(parameters.GetValueOrNull(key)) + "\"");
            }
            return header.ToString();
        }

        private void checkPreconditions(OAuthRequest request) {
            if (request.OauthParameters == null || request.OauthParameters.Count <= 0)
                throw new OAuthParametersMissingException(request);
        }
    }

}